File: Nov2020/23Nov/RotateList.cpp
/*
Problem name: 61. Rotate List
Problem link: https://leetcode.com/problems/rotate-list/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==nullptr || head->next==nullptr)    return head;
        ListNode *l1=head,*l2;
        int len=1;
        while(l1->next!=nullptr){
            len++;
            l1=l1->next;
        }
        k=len-(k%len);
        l1=head;
        while(--k){
            l1=l1->next;
        }
        l2=l1;
        while(l2->next!=nullptr){
            l2=l2->next;
        }
        l2->next=head;
        head=l1->next;
        l1->next=nullptr;
        return head;
    }
};

File: Nov2020/23Nov/RotateList.java
/*
Problem Name: 61. Rotate List
Problem Link: https://leetcode.com/problems/rotate-list/
*/
/* Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null)
            return null;
        int cnt = 0;
        ListNode temp = head, prev = null, end = null;
        while(temp != null){
            cnt++;
            end = temp;
            temp = temp.next;
        }
        k = k%cnt;
        if(k==0)
            return head;
        k = cnt - k;
        temp = head;
        while(k > 0){
            prev = temp;
            temp = temp.next;
            k--;
        }
        
        end.next = head;
        prev.next = null;
        
        return temp;
    }
}
